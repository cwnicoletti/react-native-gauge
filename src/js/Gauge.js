import React, {useRef} from 'react';
import {Animated, Platform, StyleSheet, View} from 'react-native';
import * as Progress from 'react-native-progress';
import MaskedView from '@react-native-masked-view/masked-view';
import {withAnchorPoint} from 'react-native-anchor-point';
import useDidMountEffect from '../helper/useDidMountEffect';
import {LinearGradient} from 'expo-linear-gradient';

const CIRCLE = Math.PI * 2;

const Gauge = (props) => {
  const style = StyleSheet.compose(styles.arc, props.style);

  const {
    size,
    progress,
    animated,
    overallGradient,
    addTriangleTip,
    triangleTipWidth,
    triangleTipHeight,
    triangleNeedle,
    addCircle,
    circleSize,
    addTriangleBase,
    triangleBaseWidth,
    triangleBaseHeight,
    alwaysUseEndAngle,
    endAngle,
    unfilledEndAngle,
    rotate,
    thickness,
    borderWidth,
    needleWidth,
    needleHeight,
    needleBorderRadius,
    translateNeedleY,
    color,
    borderColor,
    needleColor,
    unfilledColor,
    circleColor,
    triangleTipColor,
    triangleBaseColor,
    gradientStyle,
  } = props;

  const prevCountRef = useRef();
  useDidMountEffect(() => {
    if (animated) {
      prevCountRef.current = progress * CIRCLE * unfilledEndAngle;
      moveNeedleFn();
    }
  }, [progress, unfilledEndAngle]);

  const moveNeedle = useRef(
    new Animated.Value(progress * CIRCLE * unfilledEndAngle),
  ).current;

  const moveNeedleFn = () => {
    Animated.spring(moveNeedle, {
      toValue: progress * CIRCLE * unfilledEndAngle,
      bounciness: 0,
      useNativeDriver: true,
    }).start();
  };

  return (
    <View>
      {overallGradient ? (
        <View
          style={{
            ...styles.maskContainer,
            height: size,
            width: size,
          }}>
          <MaskedView
            maskElement={
              <Progress.Circle
                size={size}
                progress={progress}
                alwaysUseEndAngle={alwaysUseEndAngle}
                endAngle={endAngle}
                unfilledEndAngle={unfilledEndAngle}
                thickness={thickness}
                borderWidth={borderWidth}
                color={color}
                borderColor={borderColor}
                unfilledColor={unfilledColor}
                indeterminate={false}
                style={{...style, transform: [{rotate: rotate}]}}>
                <Animated.View
                  style={[
                    {
                      position: 'absolute',
                      width: triangleNeedle ? 0 : needleWidth,
                      height: triangleNeedle ? 0 : needleHeight,
                      top: size / 2 - needleHeight - translateNeedleY / 2,
                      borderTopWidth: 0,
                      borderLeftWidth: triangleNeedle ? needleWidth : 0,
                      borderRightWidth: triangleNeedle ? needleWidth : 0,
                      borderBottomWidth: triangleNeedle ? needleHeight : 0,
                      backgroundColor: triangleNeedle
                        ? 'transparent'
                        : needleColor,
                      borderRadius: triangleNeedle ? 0 : needleBorderRadius,
                      borderStyle: 'solid',
                      borderLeftColor: 'transparent',
                      borderRightColor: 'transparent',
                      borderBottomColor: needleColor,
                    },
                    withAnchorPoint(
                      {
                        transform: [
                          {rotateZ: moveNeedle},
                          {translateY: -translateNeedleY},
                        ],
                      },
                      {x: 0.5, y: 1},
                      {
                        width: needleWidth,
                        height: needleHeight + translateNeedleY,
                      },
                    ),
                  ]}>
                  {addTriangleTip && (
                    <View
                      style={{
                        alignSelf: 'center',
                        top: -triangleTipHeight,
                        borderTopWidth: 0,
                        borderStyle: 'solid',
                        borderLeftColor: 'transparent',
                        borderRightColor: 'transparent',
                        borderBottomColor: triangleTipColor,
                        borderLeftWidth: triangleTipWidth,
                        borderRightWidth: triangleTipWidth,
                        borderBottomWidth: triangleTipHeight,
                      }}
                    />
                  )}
                  {addTriangleBase && (
                    <Animated.View
                      style={{
                        position: 'absolute',
                        alignSelf: 'center',
                        top: triangleBaseHeight + needleHeight / 2 - 1,
                        borderTopWidth: 0,
                        borderStyle: 'solid',
                        borderLeftColor: 'transparent',
                        borderRightColor: 'transparent',
                        borderBottomColor: triangleBaseColor,
                        borderLeftWidth: triangleBaseWidth,
                        borderRightWidth: triangleBaseWidth,
                        borderBottomWidth: triangleBaseHeight,
                      }}
                    />
                  )}
                </Animated.View>
                {addCircle && (
                  <Animated.View
                    style={{
                      position: 'absolute',
                      height: circleSize,
                      width: circleSize,
                      borderRadius: circleSize / 2,
                      backgroundColor: circleColor,
                      transform: [{rotateZ: moveNeedle}],
                    }}
                  />
                )}
              </Progress.Circle>
            }>
            <LinearGradient
              colors={overallGradient}
              style={{
                ...gradientStyle,
                height: size,
                width: size,
              }}
            />
          </MaskedView>
        </View>
      ) : (
        <Progress.Circle
          size={size}
          progress={progress}
          alwaysUseEndAngle={alwaysUseEndAngle}
          endAngle={endAngle}
          unfilledEndAngle={unfilledEndAngle}
          thickness={thickness}
          borderWidth={borderWidth}
          color={color}
          borderColor={borderColor}
          unfilledColor={unfilledColor}
          indeterminate={false}
          style={{...style, transform: [{rotate: rotate}]}}>
          <Animated.View
            style={[
              {
                position: 'absolute',
                width: triangleNeedle ? 0 : needleWidth,
                height: triangleNeedle ? 0 : needleHeight,
                top: size / 2 - needleHeight - translateNeedleY / 2,
                borderTopWidth: 0,
                borderLeftWidth: triangleNeedle ? needleWidth : 0,
                borderRightWidth: triangleNeedle ? needleWidth : 0,
                borderBottomWidth: triangleNeedle ? needleHeight : 0,
                backgroundColor: triangleNeedle ? 'transparent' : needleColor,
                borderRadius: triangleNeedle ? 0 : needleBorderRadius,
                borderStyle: 'solid',
                borderLeftColor: 'transparent',
                borderRightColor: 'transparent',
                borderBottomColor: needleColor,
              },
              withAnchorPoint(
                {
                  transform: [
                    {rotateZ: moveNeedle},
                    {translateY: -translateNeedleY},
                  ],
                },
                {x: 0.5, y: 1},
                {
                  width: needleWidth,
                  height: needleHeight + translateNeedleY,
                },
              ),
            ]}>
            {addTriangleTip && (
              <View
                style={{
                  alignSelf: 'center',
                  top: -triangleTipHeight,
                  borderTopWidth: 0,
                  borderStyle: 'solid',
                  borderLeftColor: 'transparent',
                  borderRightColor: 'transparent',
                  borderBottomColor: triangleTipColor,
                  borderLeftWidth: triangleTipWidth,
                  borderRightWidth: triangleTipWidth,
                  borderBottomWidth: triangleTipHeight,
                }}
              />
            )}
          </Animated.View>
          {addCircle && (
            <Animated.View
              style={{
                position: 'absolute',
                height: circleSize,
                width: circleSize,
                borderRadius: circleSize / 2,
                backgroundColor: circleColor,
                transform: [{rotateZ: moveNeedle}],
              }}
            />
          )}
        </Progress.Circle>
      )}
    </View>
  );
};

Gauge.defaultProps = {
  size: 30,
  progress: 0.5,
  overallGradient: false,
  addTriangleTip: false,
  triangleTipWidth: 2,
  triangleTipHeight: 4,
  triangleNeedle: false,
  addCircle: false,
  circleSize: 15,
  addTriangleBase: false,
  triangleBaseWidth: 5,
  triangleBaseHeight: 20,
  animated: true,
  alwaysUseEndAngle: true,
  endAngle: 0.9,
  unfilledEndAngle: 0.9,
  rotate: '-90deg',
  thickness: 6,
  borderWidth: 1,
  needleWidth: 2,
  needleHeight: 45,
  needleBorderRadius: 0,
  translateNeedleY: 0,
  color: 'blue',
  borderColor: 'blue',
  needleColor: 'blue',
  unfilledColor: 'grey',
  circleColor: 'blue',
  triangleTipColor: 'blue',
  triangleBaseColor: 'blue',
  gradientStyle: {},
};

let styles;
if (Platform.OS === 'ios') {
  styles = StyleSheet.create({
    arc: {
      alignItems: 'center',
      justifyContent: 'center',
    },
    maskContainer: {
      justifyContent: 'center',
      alignItems: 'center',
    },
  });
} else {
  styles = StyleSheet.create({
    arc: {
      alignItems: 'center',
      justifyContent: 'center',
    },
    maskContainer: {
      justifyContent: 'center',
      alignItems: 'center',
    },
  });
}

export default Gauge;
