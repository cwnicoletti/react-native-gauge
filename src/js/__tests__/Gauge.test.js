import React from 'react';
import renderer from 'react-test-renderer';
import Gauge from '../Gauge';

describe('<Gauge />', () => {
  it('renders enabled gauge', () => {
    const tree = renderer.create(<Gauge />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders a gauge with custom props', () => {
    const tree = renderer.create(<Gauge size={40} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
