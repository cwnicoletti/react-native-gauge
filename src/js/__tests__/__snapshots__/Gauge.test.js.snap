// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gauge /> renders a gauge with custom props 1`] = `
<View>
  <View
    indeterminateAnimationDuration={1000}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "overflow": "hidden",
        },
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "transform": Array [
            Object {
              "rotate": "-90deg",
            },
          ],
        },
      ]
    }
  >
    <RNSVGSvgView
      bbHeight={40}
      bbWidth={40}
      collapsable={false}
      focusable={false}
      height={40}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {},
          Object {
            "flex": 0,
            "height": 40,
            "width": 40,
          },
        ]
      }
      width={40}
    >
      <RNSVGGroup>
        <RNSVGPath
          d="M24.94427190999916 35.21690426072246
          A16 16 0 0 1 10.595435963320426 7.055728090000843"
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4286611584}
          strokeLinecap={0}
          strokeWidth={6}
        />
        <RNSVGPath
          d="M20 4
          A16 16 0 0 1 24.94427190999916 35.21690426072246"
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4278190335}
          strokeLinecap={0}
          strokeWidth={6}
        />
        <RNSVGPath
          d="M20 0.5
          A19.5 19.5 0 1 1 8.53818758029677 4.224168609688526"
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4278190335}
          strokeLinecap={0}
          strokeWidth={1}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "blue",
          "borderBottomColor": "blue",
          "borderBottomWidth": 0,
          "borderLeftColor": "transparent",
          "borderLeftWidth": 0,
          "borderRadius": 0,
          "borderRightColor": "transparent",
          "borderRightWidth": 0,
          "borderStyle": "solid",
          "borderTopWidth": 0,
          "height": 45,
          "position": "absolute",
          "top": -25,
          "transform": Array [
            Object {
              "translateY": 22.5,
            },
            Object {
              "rotateZ": 2.827433388230814,
            },
            Object {
              "translateY": -0,
            },
            Object {
              "translateY": -22.5,
            },
          ],
          "width": 2,
        }
      }
    />
  </View>
</View>
`;

exports[`<Gauge /> renders enabled gauge 1`] = `
<View>
  <View
    indeterminateAnimationDuration={1000}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "overflow": "hidden",
        },
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "transform": Array [
            Object {
              "rotate": "-90deg",
            },
          ],
        },
      ]
    }
  >
    <RNSVGSvgView
      bbHeight={30}
      bbWidth={30}
      collapsable={false}
      focusable={false}
      height={30}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {},
          Object {
            "flex": 0,
            "height": 30,
            "width": 30,
          },
        ]
      }
      width={30}
    >
      <RNSVGGroup>
        <RNSVGPath
          d="M18.399186938124423 25.46162167924669
          A11 11 0 0 1 8.534362224782793 6.100813061875579"
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4286611584}
          strokeLinecap={0}
          strokeWidth={6}
        />
        <RNSVGPath
          d="M15 4
          A11 11 0 0 1 18.399186938124423 25.46162167924669"
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4278190335}
          strokeLinecap={0}
          strokeWidth={6}
        />
        <RNSVGPath
          d="M15 0.5
          A14.5 14.5 0 1 1 6.477113841759136 3.269253581563264"
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4278190335}
          strokeLinecap={0}
          strokeWidth={1}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "blue",
          "borderBottomColor": "blue",
          "borderBottomWidth": 0,
          "borderLeftColor": "transparent",
          "borderLeftWidth": 0,
          "borderRadius": 0,
          "borderRightColor": "transparent",
          "borderRightWidth": 0,
          "borderStyle": "solid",
          "borderTopWidth": 0,
          "height": 45,
          "position": "absolute",
          "top": -30,
          "transform": Array [
            Object {
              "translateY": 22.5,
            },
            Object {
              "rotateZ": 2.827433388230814,
            },
            Object {
              "translateY": -0,
            },
            Object {
              "translateY": -22.5,
            },
          ],
          "width": 2,
        }
      }
    />
  </View>
</View>
`;
